src/pybind/mgr/orchestrator.py

```diff
@@ -384,36 +386,156 @@ class DriveGroupSpec(object):
     """
     Describe a drive group in the same form that ceph-volume
     understands.
+    This structure allows the selection of storage devices
+    in a set of nodes
     """
-    def __init__(self, devices):
-        self.devices = devices
+    def _init_attribute(self, att_dict, attribute_name, default_value):
+        """ Extract attribute from dict, if not present applies default value
+
+        :param att_dict      : dictionary with attributes
+        :param attribute_name: The attribute to initialize
+        :param default_value : Value applied if the attribute is not in the dict
+        """
+
+        setattr(self,
+                attribute_name,
+                att_dict.get(attribute_name, default_value))
+
+    def _verify_device_especification(self, dev_spec):
```

Comment:
Turns out, if you convert `verify_device_especification` to a class, lots of things are going to be a bit easier:

```python
class DeviceSpecification(object):
    def __init__(self, id_model=None, size=None, rotates=None, count=None):
        """
        drive group device specification
        :param id_model: A wildcard string. e.g: "SDD*"
        :param size: Size specification of format LOW:HIGH.
            Can also take the the form :HIGH, LOW:
            or an exact value (as ceph-volume inventory reports)
        :param rotates: is the drive rotating or not
        :param count: if this is present limit the number of drives to this number.

        Any attributes (even none) can be included in the device
        specification structure.
        """
        self.id_model = id_model
        self.size = size
        self.rotates = rotates
        self.count = count

    @classmethod
    def from_json(cls, device_spec):
        return cls(**device_spec)
```

This way, you can remove all occurrences of `_verify_device_especification`, all properties

Reply:
Ok. We will use your modifications in 
https://github.com/ceph/ceph/pull/25912